lib/rmg/ObstaclePlacer.cpp

```diff
@@ -0,0 +1,187 @@
+ *
+ */
+
+#include "../mapObjects/CObjectClassesHandler.h"
+#include "ObstaclePlacer.h"
+#include "ObjectManager.h"
+#include "TreasurePlacer.h"
+#include "RockPlacer.h"
+#include "WaterRoutes.h"
+#include "WaterProxy.h"
+#include "RoadPlacer.h"
+#include "RiverPlacer.h"
+#include "RmgMap.h"
+#include "CMapGenerator.h"
+#include "../CRandomGenerator.h"
+#include "Functions.h"
+
+void ObstaclePlacer::process()
+{
+	auto * manager = zone.getModificator<ObjectManager>();
+	if(!manager)
+		return;
+	
+	auto * riverManager = zone.getModificator<RiverPlacer>();
+	
+	typedef std::vector<ObjectTemplate> obstacleVector;
+	//obstacleVector possibleObstacles;
+	
+	std::map<int, obstacleVector> obstaclesBySize;
+	typedef std::pair<int, obstacleVector> obstaclePair;
+	std::vector<obstaclePair> possibleObstacles;
+	
+	//get all possible obstacles for this terrain
+	for(auto primaryID : VLC->objtypeh->knownObjects())
+	{
+		for(auto secondaryID : VLC->objtypeh->knownSubObjects(primaryID))
+		{
+			auto handler = VLC->objtypeh->getHandlerFor(primaryID, secondaryID);
+			if(handler->isStaticObject())
+			{
+				for(auto temp : handler->getTemplates())
+				{
+					if(temp.canBePlacedAt(zone.getTerrainType()) && temp.getBlockMapOffset().valid())
+						obstaclesBySize[temp.getBlockedOffsets().size()].push_back(temp);
+				}
+			}
+		}
+	}
+	for(auto o : obstaclesBySize)
+	{
+		possibleObstacles.push_back(o);
+	}
+	boost::sort(possibleObstacles, [](const obstaclePair &p1, const obstaclePair &p2) -> bool
+	{
+		return p1.first > p2.first; //bigger obstacles first
+	});
+	
+	auto blockedArea = zone.area().getSubarea([this](const int3 & t)
+	{
+		return map.shouldBeBlocked(t);
+	});
+	blockedArea.subtract(zone.areaUsed());
+	zone.areaPossible().subtract(blockedArea);
+	
+	
+	auto prohibitedArea = zone.freePaths() + zone.areaUsed() + manager->getVisitableArea();
+	
+	//reverse order, since obstacles begin in bottom-right corner, while the map coordinates begin in top-left
+	auto blockedTiles = blockedArea.getTilesVector();
+	int tilePos = 0;
+	while(!blockedArea.empty() && tilePos < blockedArea.getTilesVector().size())
+	{
+		auto tile = blockedArea.getTilesVector()[tilePos];
+		
+		std::list<rmg::Object> allObjects;
+		std::vector<std::pair<rmg::Object*, int3>> weightedObjects; //obj + position
+		int maxWeight = std::numeric_limits<int>::min();
+		for(int i = 0; i < possibleObstacles.size(); ++i)
+		{
+			if(!possibleObstacles[i].first)
+				continue;
+			
+			for(auto & temp : possibleObstacles[i].second)
+			{
+				auto handler = VLC->objtypeh->getHandlerFor(temp.id, temp.subid);
+				auto obj = handler->create(temp);
+				allObjects.emplace_back(*obj);
+				rmg::Object * rmgObject = &allObjects.back();
+				for(auto & offset : obj->getBlockedOffsets())
+				{
+					rmgObject->setPosition(tile - offset);
+					if(!map.isOnMap(rmgObject->getPosition()))
+						continue;
+					
+					if(!rmgObject->getArea().getSubarea([this](const int3 & t)
```

Comment:
Theoretically you can have part of object to be outside of map. And it could be expected (like you may see only part of large mountain). If not, there could be negative aesthetic effect - small obstacles on the map edge only